core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -312,6 +297,18 @@ class SYCLInternal {
   int is_initialized() const { return m_queue.has_value(); }
 
   void finalize();
+
+  // fence(...) takes any type with a .wait_and_throw() method
+  // (sycl::event and sycl::queue)
+  template <typename WAT>
```

Comment:
This is template function because the implementation should be the same in both cases (`sycl::event` and `sycl::queue`).  However, because this is an unconstrained template function, I prefer having it `private` (maybe naming it `fence_helper`) And the public interface be:

`static void fence(sycl::queue& q) { fence_helper(q); }`
`static void fence(sycl::event& e) { fence_helper(e); }`

Reply:
[`5331ab0` (#3754)](https://github.com/kokkos/kokkos/pull/3754/commits/5331ab0f51f93df757d5c36c936de80701c5ebb4)